mautic/mautic
# Allow console to leverage the generic autoloader regardless of the location (part 2)

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [x]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #10957 mautic/recommended-project#7 



#### Description:

This is the successor of #10297 , and addresses the autoload issue + adds a test to checks if composer installations are not failing with the changes in a PR

This PR fixes the autoload issue, but will throw a PHP error about doctrine
```
Argument 1 passed to Doctrine\DBAL\Configuration::setResultCacheImpl()...
```
That issue will be addressed in #11053 


#### Added tests

to detect issues in the future that would break composer based installs (where the mono repo is split up, and not all files and folders are in the same place as in this repo), we want to mimic the outcome of a composer install before splitting it up.

The issue this PR fixes is one that cannot be detected without this kind of test.

This is done by using a [composer.json](https://github.com/mautic/mautic/blob/e219fd61a5782de1c6abc45f15c60215c0626abf/.github/ci-files/composer.json) that mimics the one from [mautic/recommended-project](https://github.com/mautic/recommended-project/blob/855ad15316e494150008319233a9924f1500a7c3/composer.json)

The main difference is that the `core-lib`, `plugins` and `themes` aren't pulled from packagist, but are [symlinked](https://github.com/mautic/mautic/blob/e219fd61a5782de1c6abc45f15c60215c0626abf/.github/ci-files/composer.json#L85) from the `app`, `plugins` and `themes` folders.

see e.g. the partial output of [this run](https://github.com/mautic/mautic/runs/5851982432?check_suite_focus=true):
```
...
  - Installing mautic/core-lib (4.x-dev): Symlinking from ../app
  - Installing mautic/grapes-js-builder-bundle (4.2.x-dev): Symlinking from ../plugins/GrapesJsBuilderBundle
  - Installing mautic/plugin-citrix (4.2.x-dev): Symlinking from ../plugins/MauticCitrixBundle
  - Installing mautic/plugin-clearbit (4.2.x-dev): Symlinking from ../plugins/MauticClearbitBundle
  - Installing mautic/plugin-cloudstorage (4.2.x-dev): Symlinking from ../plugins/MauticCloudStorageBundle
  - Installing mautic/plugin-crm (4.2.x-dev): Symlinking from ../plugins/MauticCrmBundle
  - Installing mautic/plugin-emailmarketing (4.2.x-dev): Symlinking from ../plugins/MauticEmailMarketingBundle
  - Installing mautic/plugin-focus (4.2.x-dev): Symlinking from ../plugins/MauticFocusBundle
  - Installing mautic/plugin-fullcontact (4.2.x-dev): Symlinking from ../plugins/MauticFullContactBundle
  - Installing mautic/plugin-gmail (4.2.x-dev): Symlinking from ../plugins/MauticGmailBundle
  - Installing mautic/plugin-outlook (4.2.x-dev): Symlinking from ../plugins/MauticOutlookBundle
  - Installing mautic/plugin-social (4.2.x-dev): Symlinking from ../plugins/MauticSocialBundle
  - Installing mautic/plugin-tagmanager (4.2.x-dev): Symlinking from ../plugins/MauticTagManagerBundle
...
```

The current test is limited to testing if the composer install works (e.g. constraint failures), and basic console commands work. This could be expanded to testing an install in the future.



#### Steps to test this PR:



* for tarball installs
  * run console commands in gitpod, they shouldn't fail
* for composer installs 
  * check the outcome of the `composer install` test
    * there is no autoload error
    * there is a doctrine error (tackled in #11053)
 



<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11039"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
